dependencies/syn/src/verus.rs

```diff
@@ -1196,22 +1258,96 @@ pub(crate) fn disallow_prefix_binop(input: crate::parse::ParseStream) -> crate::
     }
 }
 
+#[cfg(feature = "full")]
+pub(crate) fn parse_matches(
+    input: crate::parse::ParseStream,
+    lhs: Expr,
+    allow_struct: expr::parsing::AllowStruct,
+    big_and: bool,
+) -> Result<Expr> {
+    let matches_token: Token![matches] = input.parse()?;
+    let pat = input.parse()?;
+
+    let op_expr = if input.peek(Token![&&&]) {
+        if big_and {
+            let attrs = input.call(expr::parsing::expr_attrs)?;
+            let Some(rhs) = parse_prefix_binop(input, &attrs, true)? else {
+                return Err(input.error("expected &&&"))
+            };
+            Some(MatchesOpExpr {
+                op_token: MatchesOpToken::BigAnd,
+                rhs: Box::new(rhs),
+            })
+        } else {
+            return Err(input.error("in &&&, a matches expression needs to be prefixed with &&&"));
+        }
+    } else if input.peek(Token![==>]) || input.peek(Token![&&]) {
+        let op_token = if input.peek(Token![==>]) {
+            MatchesOpToken::Implies(input.parse()?)
+        } else if input.peek(Token![&&]) {
+            MatchesOpToken::AndAnd(input.parse()?)
+        } else {
+            unreachable!()
+        };
+        let mut rhs = expr::parsing::unary_expr(input, allow_struct)?;
+        loop {
+            let next = expr::parsing::peek_precedence(input);
+            if next >= expr::parsing::Precedence::Imply {
```

Comment:
For the `&&` case, this should probably be precedence And:

```
assert(E::A matches E::B && false ==> false);
```

Reply:
(Alternately, if you want `a matches b && c1 && c2 && ... && cn` to parse the `c1 && ... && cn` left associatively, you could call back into the parser recursively instead looping.)

Reply:
I preferred the former, to keep the matched variable available in further conjuncts, but let me know if you think it should be otherwise.